nds-org/workbench-helm-chart
# Support cert-manager / acmedns for issuing wildcard certs via DNS-01

## Problem
Inserting TLS certificates via jamming them into the `values.yaml` feels wrong, and this does not provide us a nice way to automate renewing them. Previously, we were unable to utilize `cert-manager` since Workbench requires wildcard certs and Google Domains does not provide an API for programmatically updating TXT records as would be required by DNS-01 to issue automated wildcard certificates.

## Approach
* Added `configHash`, to restart deployment whenever the configmap changes
* Added `workbench.ingress.tls` config options
  * `cluster_issuer`: the name of the **ClusterIssuer** to use to populate TLS secrets
  * `issuer`: the name of the (namespaced) **Issuer** to use to populate TLS secrets
* Split Ingress templates into their own file (`templates/deployment.yaml` was becoming unwieldy)
* Allow for `subdomain_prefix` to be empty

## How to Test
Prerequisites: Kubernetes 1.19+, Helm 3.3+, NGINX Ingress Controller running and accessible, ownership of a DNS domain

1. Checkout and run this branch locally
```
git clone https://github.com/nds-org/workbench-helm-chart -b support-certmgr ; cd workbench-helm-chart/ && git fetch --all && git checkout support-certmgr
```
2. Run the setup script to deploy cert-manager with `./setup-cert-manager.sh your.domain.org`
    * This will do the following:
      * Install OS dependencies
      * Deploy cert-manager Helm chart (with recursive DNS enabled - this is important)
      * Install cert-manager CRDs (e.g. `kubectl get cert`, `kubectl describe challenge`, etc)
      * Register with `auth.acme-dns.io` (not yet self-hosted):
      * Twist the resulting JSON into the format that cert-manager expects
      * Create a Kubernetes secret out of the resulting credentials
3. Create a ClusterIssuer (make sure to change the default value to YOUR email)
```
apiVersion: cert-manager.io/v1
kind: ClusterIssuer
metadata:
  name: acmedns-issuer
spec:
  acme:
    email: you@email.org
    server: https://acme-staging-v02.api.letsencrypt.org/directory
    privateKeySecretRef:
      name: acmedns-issuer-account-key
    solvers:
    - dns01:
        acmeDNS:
          # Point at the ACME-DNS you registered with
          host: https://auth.acme-dns.io
          accountSecretRef:
            # Point at the Secret that was created by the setup script
            name: acme-dns
            key: acmedns.json
```
4. Run `kubectl get clusterissuer -w` and wait for DNS to propagate (e.g. wait for issuer to be `Ready: True`)
5. Update Workbench Helm chart `values.yaml` to point at the new ClusterIssuer
    * Ensure that the following is present at the root level (not under the "workbench" key):
```yaml
certmgr:
  cluster_issuer: "acmedns-issuer"
  issuer: ""
``` 
6. Update Workbench Helm chart `values.yaml` to point at `ndslabs/apiserver:fix-apiserver-build-deps`
    * Ensure that you are using the test apiserver image (see https://github.com/nds-org/workbench-helm-chart/pull/21):
```yaml
workbench:
  images:
    apiserver: ndslabs/apiserver:fix-apiserver-build-deps
```
7. Deploy new Workbench Helm chart values to your cluster: `sudo helm upgrade --install workbench .`
8. Login into Workbench, add an application from the catalog, and check the cluster via `kubectl`
    * `kubectl get ing -n USERNAME` - This should add an Ingress resource to your cluster
    * `kubectl get ing INGRESSNAME -o yaml -n USERNAME` - Ingress should have a `cluster-issuer` annotation and `secretName` attached to it
9. Run `kubectl get cert -n USERNAME -w`
    * After a few seconds, you should see the cert go from `Ready: False` to `True`
10. Navigate to Workbench in your browser and examine the cert
    * Even though the cert is "invalid", you should see the staging certificate was issue by LE Intermediate Authority
    * NOTE: Remove "staging" from the letsencrypt URL to issue production certs
    * WARNING: LetsEncrypt's production endpoints have rate-limits on how many certificates you can issue in a given amount of time. Read up on this before changing to issuing production certs
